src/printer.rs

```diff
@@ -75,85 +93,173 @@ impl<'a> Printer<'a> {
+
+                        write!(
+                            self.handle,
+                            "{}\n{}{}",
+                            as_terminal_escaped(
+                                style,
+                                &*text,
+                                self.config.true_color,
+                                self.config.colored_output,
+                            ),
+                            " ".repeat(gutter_width),
+                            border.text.to_owned()
+                        )?;
 
+                        continue;
+                    }
+                }
+            }
+
+            write!(self.handle, "\n")?;
+        }
+
+        // Finished.
         Ok(())
     }
 
-    fn print_line_number(&self, line_number: usize) -> Option<String> {
+    /// Generates all the line decorations.
+    fn gen_decorations(&self, line_number: usize) -> Vec<PrintSegment> {
+        let mut decorations = Vec::new();
+
         if self.config.output_components.numbers() {
-            Some(
-                self.colors
-                    .line_number
-                    .paint(format!("{:4}", line_number))
-                    .to_string(),
-            )
-        } else if self.config.output_components.grid() {
-            Some("    ".to_owned())
-        } else {
-            None
+            decorations.push(self.gen_deco_line_number(line_number));
         }
-    }
 
-    fn print_git_marker(&self, line_number: usize) -> Option<String> {
         if self.config.output_components.changes() {
-            Some(
-                if let Some(ref changes) = self.line_changes {
-                    match changes.get(&(line_number as u32)) {
-                        Some(&LineChange::Added) => self.colors.git_added.paint("+"),
-                        Some(&LineChange::RemovedAbove) => self.colors.git_removed.paint("‾"),
-                        Some(&LineChange::RemovedBelow) => self.colors.git_removed.paint("_"),
-                        Some(&LineChange::Modified) => self.colors.git_modified.paint("~"),
-                        _ => Style::default().paint(" "),
-                    }
-                } else {
-                    Style::default().paint(" ")
-                }.to_string(),
-            )
-        } else if self.config.output_components.grid() {
-            Some(" ".to_owned())
-        } else {
-            None
+            decorations.push(self.gen_deco_line_changes(line_number));
         }
+
+        return decorations;
     }
 
-    fn print_line_border(&self) -> Option<String> {
-        if self.config.output_components.grid() {
-            Some(self.colors.grid.paint("│").to_string())
+    /// Generates the decoration for displaying the line number.
+    fn gen_deco_line_number(&self, line_number: usize) -> PrintSegment {
+        let plain: String = format!("{:width$}", line_number, width = LINE_NUMBER_WIDTH);
+        let color = self.colors.line_number.paint(plain.to_owned());
```

Comment:
`plain` is already a string, so `to_owned` is unnecessary

Reply:
I tried removing `to_owned()`, and ended up with this:

```
Error[E0382]: use of moved value: `plain`
   --> src/printer.rs:217:19
    |
213 |         let color = self.colors.line_number.paint(plain);
    |                                                   ----- value moved here
...
217 |             size: plain.len(),
    |                   ^^^^^ value used here after move
    |
    = note: move occurs because `plain` has type `std::string::String`, which does not implement the `Copy` trait
```

Replacing it with `plain.clone()` works, though. Would that be better to use?

Reply:
I think paint can take a `&str`, so you can pass `&plain`. also I didn't notice before, but the variable name `color` is probably not correct